patches/tModLoader/Terraria/ID/SetFactory.TML.cs

```diff
@@ -0,0 +1,159 @@
+	/// <para/> Throws an exception if the Type, data length, or default value does not match the data registered using the same key by any mod loaded before this mod.
+	/// </summary>
+	public void RegisterNamedCustomSet<T>(string key, T defaultValue, ref T[] input)
+	{
+		// Could make a ModLoader.loadStage enum or another bool, but this behaves exactly how we want anyway.
+		if (!ContentCache.contentLoadingFinished) {
+			// If a set is initialized early, throw an error if the class containing the set doesn't have ReinitializeDuringResizeArrays
+			bool willBeReinitialized = new StackTrace().GetFrames().Any(frame => frame.GetMethod()?.DeclaringType?.GetAttribute<ReinitializeDuringResizeArraysAttribute>() != null);
+			if (!willBeReinitialized)
+				throw new Exception($"Custom sets must be initialized from a class with the ReinitializeDuringResizeArrays attribute. This ensures that all content has been registered and that the custom set will have the correct length");
+		}
+
+		// TODO: Return bool to represent already exists or merged?
+
+		// Note: Intended to be load order independent as long as all parties agree on default value and Type. Any deviation will throw exception.
+
+		SetMetadata newMetadata = new SetMetadata(typeof(T), defaultValue);
+		SetMetadata existingMetadata = setMetadataMapping.GetOrAdd((key, typeof(T)), newMetadata);
+		/*if (newMetadata.type != existingMetadata.type)
+		{
+			throw new Exception($"Previously registered set for {key} is of type {existingMetadata.type} but {newMetadata.type} was supplied. Custom data set will not be registered");
+		}*/
+		if (!newMetadata.defaultValue.Equals(existingMetadata.defaultValue)) { // Primitive might be boxed, so != doesn't work.
+			throw new Exception($"Previously registered set for {key} has a default value of {existingMetadata.defaultValue} but {newMetadata.defaultValue} was supplied. Custom data set will not be registered");
+			// TODO: We could just allow this and output a warning that the data will not be shared. Keep both somehow?
+		}
+		/*if (newmetadata.length != existingmetadata.length)
+		{
+			throw new exception($"previously registered set for {key} is has length {existingmetadata.length} but supplied set has length {newmetadata.length}. custom data set will not be registered");
+		}*/
+
+		// Custom sets are registered as (SetFactory, key) in DataInstance.
+		T[] value = DataInstance<T[]>.GetOrAdd((this, key), input);
+
+		// If it already exists, merge the data
+		if (value != input) {
+			if (value.Length != input.Length) {
+				throw new Exception("Input set and existing set are of different lengths.");
+			}
+
+			// To merge, we find entries in the input that aren't defaultValue and assign them to the result.
+			// Existing changes should persist as long as mods agree on the defaultValue passed in and used in CreateXSet
+			for (int i = 0; i < input.Length; i++) {
+				if (!input[i].Equals(defaultValue)) {
```

Comment:
Standard advice is to use `EqualityComparer<T>.Default.Equals(...)` when doing generic comparisons that may include value types. Minor performance difference, and also saves you from the `NullReferenceException` you have here if `input[i] is null`